src/compiler/checker.ts

```diff
@@ -28086,15 +28086,20 @@ namespace ts {
             // type with those properties for which the binding pattern specifies a default value.
             // If the object literal is spread into another object literal, skip this step and let the top-level object
             // literal handle it instead.
-            if (contextualTypeHasPattern && node.parent.kind !== SyntaxKind.SpreadAssignment) {
-                for (const prop of getPropertiesOfType(contextualType)) {
-                    if (!propertiesTable.get(prop.escapedName) && !getPropertyOfType(spread, prop.escapedName)) {
-                        if (!(prop.flags & SymbolFlags.Optional)) {
-                            error(prop.valueDeclaration || (prop as TransientSymbol).bindingElement,
-                                Diagnostics.Initializer_provides_no_value_for_this_binding_element_and_the_binding_element_has_no_default_value);
-                        }
-                        propertiesTable.set(prop.escapedName, prop);
-                        propertiesArray.push(prop);
+            if (contextualTypeHasPattern) {
+                const rootPatternParent = findAncestor(contextualType.pattern!.parent, n => n.kind === SyntaxKind.VariableDeclaration || n.kind === SyntaxKind.BinaryExpression || n.kind === SyntaxKind.Parameter);
+                const spreadOrOutsideRootObject = findAncestor(node, n => n === rootPatternParent || (n.kind === SyntaxKind.SpreadAssignment && getContextualType(n.parent as Expression, /*contextFlags*/ undefined)!.pattern === contextualType.pattern))!;

```

Comment:
Were you able to determine that checking for the equality of these two contextual type patterns was actually necessary?

Either way, we may want to expand on the comment above to give a more rigorous definition of “If the object literal is spread into another object literal.”

Reply:
> Were you able to determine that checking for the equality of these two contextual type patterns was actually necessary?

I've removed this now since the whole test suite passes without it. It's hard for me to tell if there is any missing test case to be written that would prove that it's needed though since I don't fully understand how contextual typing works in the implementation.

> Either way, we may want to expand on the comment above to give a more rigorous definition of “If the object literal is spread into another object literal.”

The only thing that comes to my mind that could be added is adding something like "(even transitively)" to this comment.

> I agree. I don't understand how the new code implements the comment now. 

From my PoV, I didn't really change the code in a way that would require any major changes to this comment. Conceptually the previous code was just checking if the direct parent of the node is a spread assignment and now I'm checking if there is any spread assignment ancestor (up to the parent of the contextualType's parent). There are no other changes beyond that - so unless I'm missing something, or unless the comment was confusing in the first place, this change shouldn't require any major changes. The only thing that comes to my mind is what I've mentioned in the answer to Andrew's comment - I could add something like "(even transitively)". Would this make it better from your PoV? Or were you thinking about some other changes?

> And I'm worried about the number of upward tree walks in this part of the code now.

I see your point and I've tried to experiment with "context flag" and I think that I've managed to get this to work (https://github.com/microsoft/TypeScript/pull/49684/commits/be387e3bbf8a5cce2bc4c31fd77b061ea6cf8e0b). The solution isn't overly pretty but it resembles some other places in the codebase (for instance I've seen similar code with "awaited stacks" when detectecting circular `Awaited` types)

My only concern here is that now the logic depends on the top-down order of checking (which is probably a safe bet when performing full check). It was already insufficient when checking deferred nodes and that's why I had to introduce `objectsWithinSpread` set - this still relies on the "full check" being performed before checking deferred nodes (I don't quite understand why both check expressions within functions, what's the point of deferring the nodes if they are already checked during the "full check"?). So I'm somewhat worried that this might not be sufficient to check this correctly when using APIs like `getTypeAtLocation`. Thoughts?

> Also, please reformat the new code. The lines are too long.

Would it be worthwhile to introduce a related ESLint rule to the codebase? Or maybe even Prettier? :sweat_smile:

Reply:
I almost suggested setting a flag at the top of the check, but I decided against it when I realized that it’s non-trivial to decide where to _unset_ the flag—essentially, it requires the exact same logic you had looking up the tree, but split into multiple places to maintain the state (set on the way into an object literal, unset on the way into ???, unset on the way out of the object literal). And that doesn’t make it any easier to determine what "???" is. I’m not sure whether you have it correctly handled now or not. I also agree with your concern about possibly missing the context flag in calls to `getTypeAtLocation`.

I ran the perf suite after your change that included the probably-superfluous `getContextualType` addition, and it was clean. I’m pretty sure it was only calling `getContextualType` on nodes whose contextual type was already computed or was going to be computed later. Either way, I’m happy with it removed—I think we can assume it’s not necessary until proven otherwise. With that gone, I’m definitely not worried about any perf impact.

When I suggested updating the comment, I just meant for you to explain how you decide on a stopping point the same way I explained it in my first review of the PR—that it’s not enough to look at parents' `kind` since IIFEs and conditional expressions flowed the type through in a non-obvious way (this was the bug), but that we know the stopping point because we know where the binding pattern that created the contextual type is.

Reply:
@andrewbranch I've reverted the code back to the ancestor traversal variant and added more info to the discussed comment. Would you mind taking another look?